pkg/controller/flinkapplication/flink_state_machine.go

```diff
@@ -754,15 +755,52 @@ func (s *FlinkStateMachine) handleSubmittingJob(ctx context.Context, app *v1beta
 	if job == nil {
 		return statusUnchanged, errors.Errorf("Could not find job %s", s.flinkController.GetLatestJobID(ctx, app))
 	}
-
-	// wait until all vertices have been scheduled and started
-	allVerticesStarted := true
-	for _, v := range job.Vertices {
-		allVerticesStarted = allVerticesStarted && (v.StartTime > 0)
+	allVerticesRunning := true
+	if app.Spec.BetaFeaturesEnabled {
+		// wait until all vertices have been scheduled and running
+		logger.Info(ctx, "Beta features flag is enabled.")
+		jobStartTimeSec := job.StartTime / 1000
+		jobStartTimeNSec := job.StartTime % 1000
+		jobStartTime := time.Unix(jobStartTimeSec, jobStartTimeNSec).UTC()
+		now := time.Now().UTC()
+		if now.Before(jobStartTime.Add(jobVertexStateTimeoutInMinute * time.Minute)) {
+			hasFailure := false
+			failedVertexIndex := -1
+			for index, v := range job.Vertices {
+				if v.Status == client.Failed || v.Status == client.Failing {
+					failedVertexIndex = index
+					hasFailure = true
+					break
+				}
+				allVerticesRunning = allVerticesRunning && (v.StartTime > 0) && v.Status == client.Running
+			}
+			// fail fast
+			if hasFailure {
+				s.flinkController.LogEvent(ctx, app, corev1.EventTypeWarning, "JobRunningFailed",
+					fmt.Sprintf(
+						"Vertex %d with name [%s] state is Failed", failedVertexIndex, job.Vertices[failedVertexIndex].Name))
+				return s.deployFailed(app)
```

Comment:
This is a new behavior. Is it intentional ? More than monitoring, we are updating the state of the deployment here

Reply:
This change is not just monitoring but also updating state of deployment when conditions are met. 
This specific fail fast part is new and it's based on general philosophy that if fails, fail fast. 

I double checked Flink doc about the vertex state transition graph, `FAILING` could retry while `FAILED` is the final bad state. 

So I will update the logic to if any vertex state is `FAILED`, fail deployment fast.

Reply:
@anandswaminathan in retrospect this behavior did not follow the original design as we would like to move to rollingback. New PR: https://github.com/lyft/flinkk8soperator/pull/291